### PR TITLE
Implement C++ symbol demangling on macOS

### PR DESCRIPTION
**Pull request by BrettDong**
#### Summary
Infrastructure "Demangle C++ symbols in crash report on macOS"

#### Purpose of change
Try to implement C++ symbol demangling on macOS so that crash reports are easier to read.

#### Describe the solution
Extract symbol names from the output of `backtrace_symbols()` system function by regular expression and then use `abi::__cxa_demangle` in `<cxxabi.h>` to demangle C++ symbols. The pattern of the output from `backtrace_symbols()` differ across different systems, so in this PR I am only trying to implement the functionality on macOS.

#### Describe the alternative you've considered
* Instead of using `abi::__cxa_demangle` in `<cxxabi.h>`, establish a pipe and invoke `c++filt` command in the host system.
* Leave the code here as is and continue to generate raw stack trace crash logs. Let developers demangle symbol names themselves by this online tool http://demangler.com. However it does not seem to support `clang` symbols.

#### Testing
Before:
螢幕截圖 2021-02-13 下午8 46 04

After:
螢幕截圖 2021-02-13 下午8 47 25

#### Additional context
The idea was raised in #46758.

**Edit:** This should work perfectly if the binary is built natively on the same system. I don't know whether this will work in Jenkins cross-compiled version.

**Edit 2:** [Jenkins also uses clang with libc++ when cross-compiling to macOS](https://github.com/narc0tiq/catalysm-mainline-build/blob/master/mainline-build.zsh#L50) so theoretically I guess it will work too.

**Edit 3:** `<cxxabi.h>` does not exist on Windows, so I moved it to `!_WIN32` region.

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
I think this is a gcc header, so it could be available when building using Windows cross-compile/MSYS2/Cygwin (needs testing).

Demangling when using MSVC should be different.

**Comment by BrettDong** [comment]
I'll try to implement symbol demangling on macOS, Linux and BSD first, and then see if MinGW will also work. I don't have a x86 arch computer right now so I'm unable to code and test MSVC demangling.

**Comment by BrettDong** [comment]
I tried to install Visual Studio 2019 on Windows 10 ARM VM but the debugger doesn't work.

File: src/debug.cpp [hunk 1]
```diff
@@ -791,6 +793,34 @@ static constexpr int bt_cnt = 20;
 static void *bt[bt_cnt];
 #endif
 
+#if !defined(_WIN32)
+static void write_demangled_frame( std::ostream &out, const char *frame )
```
Comment (ZhilkinSerg) [inline src/debug.cpp:797]: We need to hide this function for Android:

```
20:11:27 [arm64-v8a] Compile++      : main <= debug.cpp

20:11:27 /var/lib/jenkins/workspace/Cataclysm-Android/android/app/jni/src/../../../../src/debug.cpp:831:13: warning: unused function 'write_demangled_frame' [-Wunused-function]
20:11:27 static void write_demangled_frame( std::ostream &out, const char *frame )
20:11:27             ^
20:11:27 1 warning generated.
```

**Pull request merged**